waspc/src/Wasp/AppSpec/Job.hs

```diff
@@ -26,38 +30,41 @@ data Job = Job
     schedule :: Maybe Schedule,
     entities :: Maybe [Ref Entity]
   }
-  deriving (Show, Eq, Data)
+  deriving (Show, Eq, Data, Generic, FromJSON)
 
 instance IsDecl Job
 
 data JobExecutor = PgBoss
-  deriving (Show, Eq, Data, Ord, Enum, Bounded)
+  deriving (Show, Eq, Data, Ord, Enum, Bounded, Generic)
+
+-- NOTE: For some reason, deriving FromJSON for JobExecutor does not work. I'm
+-- guessing it's because "PgBoss" is the only data constructor (the same code
+-- works as expected with EmailProvider, which is very similar).
```

Comment:
Hah that is super weird! No idea why that wouldn't work. Have you tried deriving ToJSON instance for it and then checking out what it generates?

Ok I actually give it a quick try in GHCI:
![image](https://github.com/user-attachments/assets/ff46953c-df55-4197-a553-6bf30f648dd4)

So for `data Foo = Bar`, it expects `[]`! Ha that is weird. Is it a bug in Aeson? Or do they have some reason for this? Weird.

Reply:
I tried it out. It's not a bug. Aeson uses different representations for different kinds of data constructors.

## Case 1
In the most general case (types with two or more data constructors and one or more fields per data constructor), it explicitly lists the tag under `tag` and the fields under `contents` (as a heterogenous array):
```
ghci> data Foo = Bar String Int | Baz Int deriving (Generic, FromJSON, ToJSON, Show)
ghci> encoded = encode $ Bar "marko" 12
ghci> encoded
"{\"contents\":[\"marko\",12],\"tag\":\"Bar\"}"
ghci> decode encoded :: Maybe Foo
Just (Bar "marko" 12)
```

## Case 2
For types with two or more data constructors and no fields, it "saves" space by just encoding the tags (without explicit `tag` and `contents` labels, since `contents` would always be empty):
```
ghci> data Foo = Bar | Baz deriving (Generic, FromJSON, ToJSON, Show)
ghci> encoded = encode $ Bar
ghci> encoded
"\"Bar\""
ghci> decode encoded :: Maybe Foo
Just Bar
```

## Case 3
For types with a single data constructor and two or more fields, it just outputs a heterogenous array:
```
ghci> data Foo = Bar String Int deriving (Generic, FromJSON, ToJSON, Show)
ghci> encoded = encode $ Bar "marko" 12
ghci> encoded
"[\"marko\",12]"
ghci> decode encoded :: Maybe Foo
Just (Bar "marko" 12)
```

## Case 4
I guess this is why, for types that only feature a single data constructor with no fields, it outputs only an empty array (it's like an even more special case of Case 3)
```
ghci>  data Foo = Bar deriving (Generic, FromJSON, ToJSON, Show)
ghci> encoded = encode Bar
ghci> encoded
"[]"
ghci> decode encoded :: Maybe Foo
Just Bar
```

## Conclusion
I guess it always tries to choose the minimal unambiguous representation of a type. It may be unintuitive for humans, but it's space-efficient (which might matter with JSON).

Reply:
Ok interesting, thanks for analysis. Still a bit shitty, not a default one would expect, but yeah I guess they optimized for space usage, not minimal changes if type definition changes.